### PR TITLE
[CSL-716] Add UnknownAddress

### PR DESCRIPTION
**Pull request by georgeee**

### CODE REVIEW

File: src/Pos/Types/Types.hs [hunk 1]
```diff
@@ -95,6 +96,7 @@ data TxInWitness
                 , twSig :: TxSig}
     | ScriptWitness { twValidator :: Script
                     , twRedeemer  :: Script}
+    | UnknownWitnessType Word8 ByteString
```
Comment (georgeee) [inline src/Pos/Types/Types.hs]: Minor: types should be with `!`

**Comment by neongreen** [comment]
> If tx is received via Inv/Req/Data, it's rejected if it has address/witness/script of unknown type

Yes, my “Done.” wasn't exactly accurate. (Sorry!) I meant that I was done with modifying the types and wanted your feedback before proceeding to modify message handlers.

File: src/Pos/Types/Types.hs (hunk 1 above)
Comment (neongreen) [inline src/Pos/Types/Types.hs]: I'll add `!` to all fields here, then

**Comment by neongreen** [comment]
Updated. 

(A minor question: is comparing block versions with `>=` okay or should I write a separate comparison operator?)

**Comment by georgeee** [comment]
@neongreen not sure. `1.1.0` is not `== 1.1.1`, neither it's `>` nor `<`. There is linear order only on `maj.min` part of block version

File: src/Pos/Types/Block/Functions.hs [hunk 2]
```diff
@@ -496,6 +509,47 @@ verifyBlock VerifyBlockParams {..} blk =
               , "Some of block's PSKs have the same issuer, which is prohibited")
             ]
         | otherwise = mempty
+    verifyVersions bv = case blk of
+        Right mainBlk | bv >= mainBlk ^. gbHeader . gbhExtra . mehBlockVersion ->
```
Comment (georgeee) [inline src/Pos/Types/Block/Functions.hs]: Here should be condition:
`lastKnownBlockVersion >= min (lastAdoptedBV, mainBlk ^. gbHeader . gbhExtra . mehBlockVersion)`

File: src/Pos/Types/Block/Functions.hs [hunk 3]
```diff
@@ -515,9 +569,10 @@ verifyBlocks
        )
     => Maybe SlotId
     -> Maybe SlotLeaders
+    -> Maybe BlockVersion          -- ^ “Effective block version”
```
Comment (georgeee) [inline src/Pos/Types/Block/Functions.hs]: No. Effective version is `min(adoptedBV, block.version)`

File: src/Pos/Types/Block/Functions.hs (hunk 2 above)
Comment (gromakovsky) [inline src/Pos/Types/Block/Functions.hs]: If `mainBlk ^. gbHeader . gbhExtra . mehBlockVersion` is less than `lastAdoptedBV`, such block is rejected.

File: src/Pos/Types/Block/Functions.hs (hunk 3 above)
Comment (gromakovsky) [inline src/Pos/Types/Block/Functions.hs]: We don't accept blocks with versions less than adopted BV.

File: src/Pos/Types/Block/Functions.hs (hunk 2 above)
Comment (georgeee) [inline src/Pos/Types/Block/Functions.hs]: Then simply `lastKnownBlockVersion >= lastAdoptedBV`

File: src/Pos/Types/Block/Functions.hs (hunk 3 above)
Comment (georgeee) [inline src/Pos/Types/Block/Functions.hs]: Then it's just last adopted version, get rid of "effective"

**Comment by neongreen** [comment]
Merged manually.

File: src/Pos/Aeson/Types.hs [hunk 4]
```diff
@@ -4,22 +4,37 @@ module Pos.Aeson.Types
...
+    toJSON PubKeyAddress{..} = object [
+        "tag" .= ("PubKeyAddress" :: Text),
+        "addrKeyHash" .= addrKeyHash ]
+    toJSON ScriptAddress{..} = object [
+        "tag" .= ("ScriptAddress" :: Text),
+        "addrScriptHash" .= addrScriptHash ]
+    toJSON (UnknownAddressType t bs) = object [
+        "tag" .= ("UnknownAddressType" :: Text),
+        "type" .= t,
+        "content" .= sformat base16F bs ]
```
Comment (akegalj) [inline src/Pos/Aeson/Types.hs]: Tested, everything works fine!

**Pull request merged**